clang/include/clang/ExtractAPI/Serialization/SerializerBase.h

```diff
@@ -28,11 +28,78 @@ struct APISerializerOption {
 };
 
 /// The base interface of serializers for API information.
-class APISerializer {
+template <typename Derived> class APISerializer {
 public:
   /// Serialize the API information to \p os.
   virtual void serialize(raw_ostream &os) = 0;
 
+  virtual void traverseGlobalFunctionRecords() {
```

Comment:
The nice thing about using the [CRTP](https://en.cppreference.com/w/cpp/language/crtp) (the thing with `template <typename Derived>` is that you don't need to mark the methods as virtual. You lose [dynamic dispatch](https://condor.depaul.edu/ichu/csc447/notes/wk10/Dynamic2.htm) but this allows you to statically dispatch  the calls which is really the behavior we want here.